ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -263,36 +322,73 @@ txSubmissionInbound _tracer maxUnacked TxSubmissionMempoolWriter{..} =
     serverReqTxs :: forall (n :: N).
                     Nat n
                  -> ServerState txid tx
-                 -> ServerStIdle n txid tx m ()
-    serverReqTxs n st =
-        SendMsgRequestTxsPipelined
+                 -> m (ServerStIdle n txid tx m ())
+    serverReqTxs n st = do
+        -- We can pick txs out of order if we want to, and we can decide not
+        -- to pick some txs at all (if for example we have seen them already).
+        -- We don't have to pick them all in one go, we can stick to a limit
+        -- like maxTxToRequest.
+
+        -- TODO @intricate: Should this whole thing be an STM tx?
+        recentTxIds <- atomically $ readTVar recentTxIdsVar
```

Comment:
How so? This `readTVar` is the only thing that needs `STM`, the remainder of this function is pure.